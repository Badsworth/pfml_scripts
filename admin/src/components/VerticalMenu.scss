.menu {
  position: relative;
}

.menu__icon {
  background: none;
  border: none;
  cursor: pointer;
  font-size: 1.14em;
  position: relative;
  vertical-align: middle;
}

.menu__dropdown {
  background-color: white;
  border-radius: 6px;
  box-shadow: $Box-Shadow;
  cursor: pointer;
  left: -100px;
  margin-top: 0;
  padding: 1.14em 1.14em 0;
  position: absolute;
  user-select: none;
  width: 200px;
}

.menu__item {
  list-style: none;
  margin-bottom: 1em;
}

.menu__item--disabled {
  color: $Color-GraniteGray-300;
}
