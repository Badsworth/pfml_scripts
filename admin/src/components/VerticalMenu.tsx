import { useState, isValidElement } from "react";
import Link from "next/link";

type option = {
  enabled: boolean;
  text: string;
  type: "button" | "link";
  onClick?: Function;
  href?: string;
};

type Props = {
  options: option[];
};

export default function VerticalMenu({ options }: Props) {
  const [open, setOpen] = useState(false);

  const toggleOpen = () => {
    setOpen(!open);
  };

  const handleOnClick = (event: React.MouseEvent, onClick?: Function) => {
    event.preventDefault();

    onClick && onClick(event);
  };

  const listItems = options.map((option: option) => {
    if (!option.enabled) {
      return (
        <span className="vertical-menu__item vertical-menu__item--disabled">
          {option.text}
        </span>
      );
    }
    if (option.type === "link" && option.href) {
      return (
        <Link href={option.href}>
          <a className="vertical-menu__item vertical-menu__link">
            {option.text}
          </a>
        </Link>
      );
    }
    if (option.type === "button" && option.onClick) {
      return (
        <button
          className="vertical-menu__item vertical-menu__button"
          onClick={(e) => {
            handleOnClick(e, option.onClick);
          }}
          type="button"
        >
          {option.text}
        </button>
      );
    }
    return null;
  });

  return (
    <div className="vertical-menu">
      <button
        className="vertical-menu__icon pfml-icon--vmenu"
        onClick={toggleOpen}
        type="button"
      ></button>

      {open && (
        <ul className="vertical-menu__dropdown">
          {listItems.map((listItem, index) => {
            return isValidElement(listItem) && listItem.type === "span" ? (
              <li key={index}>{listItem}</li>
            ) : (
              <li key={index} onClick={toggleOpen}>
                {listItem}
              </li>
            );
          })}
        </ul>
      )}
    </div>
  );
}
