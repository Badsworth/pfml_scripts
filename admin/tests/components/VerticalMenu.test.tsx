import React from "react";
import { create } from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import VerticalMenu, {
  Option as VerticalMenuOption,
  Props as VerticalMenuProps,
} from "../../src/components/VerticalMenu";

// Make all optional for passed props so we
// only have to pass what we want to test
type PassedProps = {
  options?: VerticalMenuOption[];
};

const renderComponent = (passedProps: PassedProps = {}) => {
  // Prop defaults
  const props: VerticalMenuProps = {
    options: [
      {
        enabled: true,
        text: "Test Option",
        type: "button",
      },
    ],
    // Overwrite defaults with passed props
    ...passedProps,
  };

  return render(<VerticalMenu {...props} />);
};

describe("VerticalMenu", () => {
  test("renders the default component", () => {
    const props: VerticalMenuProps = {
      options: [
        {
          enabled: true,
          text: "Test Option",
          type: "button",
        },
      ],
    };

    const component = create(<VerticalMenu {...props} />).toJSON();
    expect(component).toMatchSnapshot();
  });

  test("renders menu with enabled buttons", () => {
    renderComponent({
      options: [
        {
          enabled: true,
          text: "Button One",
          type: "button",
          onClick: jest.fn(),
        },
        {
          enabled: true,
          text: "Button Two",
          type: "button",
          onClick: jest.fn(),
        },
      ],
    });

    userEvent.click(screen.getByTestId("vertical-menu-trigger"));

    expect(
      screen.getByRole("button", { name: "Button One" }),
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: "Button Two" }),
    ).toBeInTheDocument();
  });

  test("renders menu with disabled buttons", () => {
    renderComponent({
      options: [
        {
          enabled: false,
          text: "Button One",
          type: "button",
          onClick: jest.fn(),
        },
        {
          enabled: false,
          text: "Button Two",
          type: "button",
          onClick: jest.fn(),
        },
      ],
    });

    userEvent.click(screen.getByTestId("vertical-menu-trigger"));

    expect(
      screen.queryByRole("button", { name: "Button One" }),
    ).not.toBeInTheDocument();
    expect(
      screen.queryByRole("button", { name: "Button Two" }),
    ).not.toBeInTheDocument();

    expect(screen.getByText("Button One")).toBeInTheDocument();
    expect(screen.getByText("Button Two")).toBeInTheDocument();
  });

  test("renders menu with enabled links", () => {
    renderComponent({
      options: [
        {
          enabled: true,
          text: "Link One",
          type: "link",
          href: "#",
        },
        {
          enabled: true,
          text: "Link Two",
          type: "link",
          href: "#",
        },
      ],
    });

    userEvent.click(screen.getByTestId("vertical-menu-trigger"));

    expect(screen.getByRole("link", { name: "Link One" })).toBeInTheDocument();
    expect(screen.getByRole("link", { name: "Link Two" })).toBeInTheDocument();
  });

  test("renders menu with disabled links", () => {
    renderComponent({
      options: [
        {
          enabled: false,
          text: "Link One",
          type: "link",
          href: "#",
        },
        {
          enabled: false,
          text: "Link Two",
          type: "link",
          href: "#",
        },
      ],
    });

    userEvent.click(screen.getByTestId("vertical-menu-trigger"));

    expect(
      screen.queryByRole("link", { name: "Link One" }),
    ).not.toBeInTheDocument();
    expect(
      screen.queryByRole("link", { name: "Link Two" }),
    ).not.toBeInTheDocument();

    expect(screen.getByText("Link One")).toBeInTheDocument();
    expect(screen.getByText("Link Two")).toBeInTheDocument();
  });

  test("fires `onClick` callback prop when button is clicked", () => {
    const onClick = jest.fn();

    renderComponent({
      options: [
        {
          enabled: true,
          text: "Button One",
          type: "button",
          onClick: onClick,
        },
        {
          enabled: true,
          text: "Button Two",
          type: "button",
          onClick: jest.fn(),
        },
      ],
    });

    userEvent.click(screen.getByTestId("vertical-menu-trigger"));
    userEvent.click(screen.getByRole("button", { name: "Button One" }));

    expect(onClick).toHaveBeenCalled();
  });
});
