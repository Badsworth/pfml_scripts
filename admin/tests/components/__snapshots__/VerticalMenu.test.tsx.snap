// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerticalMenu renders the default component 1`] = `
<div
  className="vertical-menu"
>
  <button
    className="vertical-menu-trigger"
    data-testid="vertical-menu-trigger"
    onClick={[Function]}
    type="button"
  >
    <svg
      aria-hidden="true"
      className="vertical-menu-trigger__icon"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10 6a2 2 0 110-4 2 2 0 010 4zM10 12a2 2 0 110-4 2 2 0 010 4zM10 18a2 2 0 110-4 2 2 0 010 4z"
      />
    </svg>
  </button>
</div>
`;
